isinstance can coerce a known type to become None

Given the following code (mypy 0.590 and python 3.6.5):
```
from typing import *

class A(object):
    def method(self) -> str:
        return ''

R = TypeVar('R')

class B(object):
    def func(self, instance: A, instance_type: Type[R]) -> None:
        reveal_type(instance)
        if isinstance(instance, instance_type):
            reveal_type(instance)
            instance.method()
```
I get the following response:
```
test.py:11: error: Revealed type is 'test.A'
test.py:13: error: Revealed type is 'builtins.None'
test.py:14: error: "None" has no attribute "method"
```
I would have expected the type of instance to have at the very least been a subclass of A, given that it was before it went into the call.  So something like:
```
test.py:11: error: Revealed type is 'test.A'
test.py:13: error: Revealed type is 'R'
```
Or perhaps more complex to indicate it's subclasses from both A and R, (but I don't know the syntax for that), but I would not expect to see the `"None" has no attribute "method"` error.  It makes no difference whether R is bound to A or not.

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -173,6 +173,15 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
         return declared.copy_modified(
             upper_bound=narrow_declared_type(declared.upper_bound, original_narrowed)
         )
+    elif (
+        isinstance(narrowed, TypeVarType)
+        and not has_type_vars(original_declared)
+        and is_subtype(original_declared, narrowed.upper_bound)
+    ):
+        # This branch is a mirror image of the above one.
+        return narrowed.copy_modified(
+            upper_bound=narrow_declared_type(original_declared, narrowed.upper_bound)
+        )
     elif not is_overlapping_types(declared, narrowed):
         if state.strict_optional:
             return UninhabitedType()

--- a/test-data/unit/check-narrowing.test
+++ b/test-data/unit/check-narrowing.test
@@ -4113,3 +4113,25 @@ def stringify(value: A | B | C) -> str:
         return value.name
     return ""
 [builtins fixtures/tuple.pyi]
+
+[case testTypeVariableBoundNarrowingSymmetric]
+from typing import TypeVar
+
+class B: ...
+class C(B):
+    x: int
+
+H = TypeVar('H', bound=B)
+
+def func(y: H) -> H:
+    x: C
+    assert x == y
+    reveal_type(x)  # N: Revealed type is "H`-1"
+    reveal_type(y)  # N: Revealed type is "H`-1"
+    reveal_type(x.x)  # N: Revealed type is "builtins.int"
+    reveal_type(y.x)  # N: Revealed type is "builtins.int"
+    if bool():
+        return x
+    else:
+        return y
+[builtins fixtures/primitives.pyi]

